src/consensus/aft/test/committable_suffix.cpp

```diff
@@ -0,0 +1,865 @@
+    auto msg_type = serialized::peek<aft::RaftMsgType>(data, size);
+    if (msg_type == aft::raft_append_entries)
+    {
+      const auto ae = *(aft::AppendEntries*)data;
+
+      const auto best_it = best.find(target);
+      if (best_it == best.end() || best_it->second.first > ae.prev_idx)
+      {
+        best[target] = std::make_pair(ae.prev_idx, it);
+      }
+    }
+  }
+
+  aft::ChannelStubProxy::MessageList best_only;
+  for (const auto& [node_id, pair] : best)
+  {
+    const auto& [idx, it] = pair;
+    best_only.push_back(*it);
+  }
+
+  messages = std::move(best_only);
+}
+
+#define TEST_DECLARE_NODE(N) \
+  ccf::NodeId node_id##N(#N); \
+  auto store##N = std::make_shared<AllSigsStore>(node_id##N); \
+  TRaft r##N( \
+    ConsensusType::CFT, \
+    std::make_unique<AllSigsAdaptor>(store##N), \
+    std::make_unique<aft::LedgerStubProxy>(node_id##N), \
+    std::make_shared<aft::ChannelStubProxy>(), \
+    std::make_shared<aft::StubSnapshotter>(), \
+    nullptr, \
+    nullptr, \
+    cert, \
+    std::make_shared<aft::State>(node_id##N), \
+    nullptr, \
+    nullptr, \
+    nullptr, \
+    request_timeout, \
+    election_timeout, \
+    election_timeout); \
+  initial_config[node_id##N] = {}; \
+  nodes[node_id##N] = &r##N; \
+  auto channels##N = channel_stub_proxy(r##N);
+
+/**
+  Summary of the behaviour this test aims to explore (for implementation reasons
+  the indices in the actual test don't match those in this comment, but the
+  logical relationships are the same):
+
+  - Produce an initial state where A is primary in term 1 of a 5-node network,
+    has mixed success replicating its entries. Each node's commit index is
+    marked by []. The highest index known on f+1 nodes must be persisted, and
+    when enough ACKs are received this becomes the commit index on the primary.
+    In this case, that index _is_ known as the commit index on the, which is
+    [1.2].
+    A:  1.1 [1.2] 1.3
+    B: [1.1] 1.2  1.3
+    C: [1.1] 1.2
+    D: [1.1]
+    E: [1.1]
+
+  - Intuitively, B and C are responsible for persisting 1.2, although they don't
+    know this locally.
+
+  - Node A dies. Node B becomes primary, and creates some tail junk that no-one
+    else sees. It makes no commit progress. Crucially this is committable (as
+    all of these indices are), so B is reluctant to discard the entries in term
+    2, and 1.3, though all _could_ be discarded.
+    A:  1.1 [1.2] 1.3 (DEAD)
+    B: [1.1] 1.2  1.3  2.4  2.5
+    C: [1.1] 1.2
+    D: [1.1]
+    E: [1.1]
+
+  - Node C doesn't hear from B for long enough, and becomes primary. Its
+    AppendEntries are regularly lost. Eventually a heartbeat ("I'm at 1.2")
+    reaches Node B, which responds with a NACK ("I don't accept that tail").
+    Node C responds to that NACK with a better AppendEntries, which _should_
+    bring Node B back to the committed level, and forward to wherever C is.
+    However, under the current implementation, this AppendEntries may
```

Comment:
"under the current implementation" will be outdated once this is merged. Should we briefly mention what the first implementation was instead?

Reply:
Good point, I've updated this comment.

```
However, if the AppendEntries returns an early-estimate of matching index
(as a previous implementation did), this AppendEntries may actually cause B
to roll back further than is safe, losing the committed state.
```